rust/ranked-prolly-tree/README.md

```diff
@@ -0,0 +1,153 @@
+# ranked-prolly-tree
+
+
+`ranked-prolly-tree` (RPT) is a key-value store implemented as a [prolly tree], utilizing a flexible content-addressed block storage backend. RPT is designed to be the foundation of a lazy database, utilizing partial sync on demand.
+
+## Usage
+
+```rust
+use ranked_prolly_tree::{Tree, Result};
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let mut tree = Tree::<64, _>::default();
+    let hello = String::from("hello").into_bytes();
+    let world = String::from("world").into_bytes();
+    tree.set(hello.clone(), world.clone()).await?;
+    assert_eq!(tree.get(&hello).await?.unwrap(), world);
+    Ok(())
+}
+```
+
+## Support
+
+|                       | native | wasm32 browser |
+| :-------------------- | -----: | -------------: |
+| [`MemoryStore`]       |   ✅   |       ✅       |
+| [`SyncMemoryStore`]   |   ✅   |       ✅       |
+| [`FileSystemStore`]   |   ✅   |       ❌       |
+| [`IndexedDbStore`]    |   ❌   |       ✅       |
+
+## Design
+
+RPT is the basis for a lazy database that can efficiently synchronize with a remote block store. Repeatedly requesting data from a server incurs overhead, and we attempt to return addition data in blocks to reduce latency in future requests with caching, especially when nearby groups of keys are accessed.
+
+RPT is a [prolly tree] implementation where nodes are represented by a *block* in a block store. The bottom level contains **leaf** (💚) nodes, which represent key/value pairs, ordered by their keys, and are always at the same level. **Segments** (💙) are parents of leaf nodes. **Branches** (💜) are nodes whose children are either segments or other branches.
+
+An example tree of 5 entries (with keys 1, 2, 7, 8, 9, values omitted) and their rank, and segment and branches with the maximum key, or the largest/right-most key in their descendants:
```

Comment:
I think I have mentioned this in one of our calls, but it may have gotten lost. I think it may be a good idea to capture key ranges as opposed to just upper bound. This would make it possible to abort early when looking for key or scanning out of range, specifically it will let us know gaps between nodes and consequently identify whether specific key (or range of keys) is even contained by the tree.

Also when encoding adjacent nodes (when there are no gaps) we could leave out lower bound and derive it from the node on the left.

It may prove useful in some access patterns and impose overhead in the others, so ideally it would be an option al feature. From these perspective range with no lower bound can be treated as probabilistically containing key range spanning from left node upper bound, while presence of lower bound would guarantee the presence of key range.

Reply:
Filed https://github.com/commontoolsinc/system/issues/273